### DISCUSSION TITLE
Semantic Model Serialization

### DISCUSSION THREAD

**Discussion by Lotes** [eclipse-langium/langium#566] [Ideas]
I had some thoughts about migrating source codes between versions Vn and Vn+1. Or also transpiling language A to B.

For both ideas it would be helpful to have a conversion of the AST back to source code.

Especially migrations would be cool 😎. If I imagine that we would have an incremental or agile approach, it could be helpful for developers to have the possibility to perform upgrades or even downgrades.

WDYT? Are there maybe additional usecases?

**Comment by msujew** [comment]
@Lotes We had this discussion some time ago (see Slack), and decided to either not include it into Langium at all (and let users implement a domain generator instead), or postpone it until there's an incredibly alluring use case. There are just too many pitfalls in there.

**Reply by spoenemann** [reply]
I agree this is too complex and low prio for now, but the topic is still interesting! I propose that we kick-off the discussion after a v1.0.0 release. I could imagine implementing this in a separate package (not in the `langium` core package).

**Comment by RobertWalterU3d** [comment]
IIUC (and I might not), what's discussed here is the idea of an "Unparser". One additional use case where this would come in handy is when you want to support non-textual editors for a language as well. We've prototyped that some time ago where users could edit a program both as text (a Xtext DSL) and via a proprietary visual scripting editor.  
I see there are plans to integrate Langium and Sprotty. Is that meant in a way where Sprotty is purely used as a view (i.e. not an editor)?

**Reply by msujew** [reply]
The way the Sprotty integration works for now, is as a read-only view. See the [langium-sprotty](https://github.com/langium/langium/tree/main/packages/langium-sprotty) package. There's nothing really that stops anyone from extending this, but it's not our own priority.
